### PR TITLE
Google Sign-On with new libraries

### LINKED ISSUE #5776: OpenEMR Google Sign-On does not work
### Describe the bug

After installing OpenEMR 7  when I try to login by using Google Sign on button it takes me to the same login page. Nothing happens!
Console log shows the following error
"You have created a new client application that uses libraries for user authentication or authorization that will soon be deprecated. New clients must use the new libraries instead; existing clients must also migrate before these libraries are deprecated. See the Migration Guide for more information."

### To Reproduce

1. Add a '....'

### Expected behavior

Screen Shot 2022-09-19 at 6 55 22 AM

### Client configuration

- Browser: Chrome
- OpenEMR version: 7
- Operating system: Linux

### PR DESCRIPTION
**Pull request by Saparbek-Nagashibekov**
Fixes #5776

#### Short description of what this resolves: 
OpenEMR Google Sign-On with new libraries

#### Changes proposed in this pull request:
Added Google new library changes

### CODE REVIEW

File: templates/login/login_core.html.twig [hunk 1]
```diff
@@ -1,14 +1,16 @@
-{% set srOnly = (showLabels == true) ? "" : "sr-only" %}
 <!DOCTYPE html>
 <html lang="en">
 <head>
-    {% if displayGoogleSignin %}
-        <meta name="google-signin-client_id" content="{{ googleSigninClientID|attr }}">
-    {% endif %}
+   <script src="https://accounts.google.com/gsi/client" async defer></script>
+    <div id="g_id_onload"
+         data-client_id="{{ googleSigninClientID }}"
```
Comment (stephenwaite) [inline templates/login/login_core.html.twig:6]: hi @Saparbek-Nagashibekov, thanks for the PR, think this needs to stay as `googleSigninClientID|attr`

File: templates/login/login_core.html.twig [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
+         data-callback="onSignInSuccess" data-auto_prompt="false"> 
+   </div>
+ 
     <title>{{ title|text }} {{ "Login"|xlt }}</title>
     {% block head %}{% endblock %}
-    <script src="{{ webroot|attr }}/interface/product_registration/product_registration_service.js?v={{ assetVersion|attr_url }}"></script>
-    <script src="{{ webroot|attr }}/interface/product_registration/product_registration_controller.js?v={{ assetVersion|attr_url }}"></script>
+    <script src="{{ webroot|attr }}/interface/product_registration/product_registration_service.js?v={{ assetVersion|attr }}"></script>
```
Comment (stephenwaite) [inline templates/login/login_core.html.twig:12]: and these as `attr|url`

File: templates/login/login_core.html.twig [hunk 3]
```diff
@@ -203,23 +206,23 @@
...
     }
 
     // When Google sign-in successful, sign in to the app, but only
     // if the button was clicked (otherwise we would automatically login)
     function onSignInSuccess(googleUser) {
-        if (google_signin === true) {
-            const auth_response = googleUser.getAuthResponse();
-            const id_token = auth_response.id_token;
+        
+    //        const auth_response = googleUser.getAuthResponse();
```
Comment (stephenwaite) [inline templates/login/login_core.html.twig:216]: ok to remove this instead of commenting

File: templates/login/login_core.html.twig [hunk 4]
```diff
@@ -249,19 +252,13 @@
             $('#standard-auth-username, #standard-auth-password').show();
         });
     }
-
-    $.getScript('https://apis.google.com/js/platform.js', function (data, textStatus, jqxh) {
+/*
```
Comment (stephenwaite) [inline templates/login/login_core.html.twig:255]: ok to remove

File: templates/login/login_core.html.twig (hunk 3 above)
Comment (Saparbek-Nagashibekov) [reply templates/login/login_core.html.twig]: @stephenwaite I made changes like you said. Could you review 
commit cd882f845a5caab6d61894e474603a5dbd362171  ?
What is the next step?
Thanks

File: templates/login/login_core.html.twig [hunk 5]
```diff
@@ -1,10 +1,15 @@
-{% set srOnly = (showLabels == true) ? "" : "sr-only" %}
 <!DOCTYPE html>
 <html lang="en">
 <head>
     {% if displayGoogleSignin %}
         <meta name="google-signin-client_id" content="{{ googleSigninClientID|attr }}">
     {% endif %}
+   <script src="https://accounts.google.com/gsi/client" async defer></script>
+    <div id="g_id_onload"
+         data-client_id="{{ googleSigninClientID }}"
```
Comment (bradymiller) [inline templates/login/login_core.html.twig:9]: need to html escape this via attr:
```
data-client_id="{{ googleSigninClientID|attr }}"
```

File: templates/login/login_core.html.twig [hunk 6]
```diff
@@ -1,10 +1,15 @@
...
+         data-callback="onSignInSuccess" data-auto_prompt="false"> 
+   </div>
```
Comment (bradymiller) [inline templates/login/login_core.html.twig:11]: shouldn't this block of code be within the above if conditional in line 4?
(ie. only shows `if displayGoogleSignin`)

File: templates/login/login_core.html.twig [hunk 7]
```diff
@@ -1,10 +1,15 @@
-{% set srOnly = (showLabels == true) ? "" : "sr-only" %}
```
Comment (bradymiller) [inline templates/login/login_core.html.twig:1]: above line looks like an unrelated change. If so, would add it back.

**Comment by bradymiller** [comment]
hi @Saparbek-Nagashibekov , thanks for addressing this bug! Just noted a couple issues. 
Quick question:
It looks like this new library is required for new users (per the forum). Will these changes be backward compatible for users that are using the old library?

File: templates/login/login_core.html.twig (hunk 5 above)
Comment (Saparbek-Nagashibekov) [reply templates/login/login_core.html.twig:9]: Hi, how to do that ? could you give your suggested change? I will add it. Thanks

Comment (Saparbek-Nagashibekov) [reply templates/login/login_core.html.twig:9]: @bradymiller

Comment (bradymiller) [reply templates/login/login_core.html.twig:9]: i should of been more clear. you pipe it through the attr to html escape it there:
```
data-client_id="{{ googleSigninClientID|attr }}"
```

Comment (Saparbek-Nagashibekov) [reply templates/login/login_core.html.twig:9]: What does it mean "html escape this via attr" how can I do that? Could you give me sample. thanks

Comment (bradymiller) [reply templates/login/login_core.html.twig:9]: here is what the line should look like example (see where the `|attr` is in below line :) ):
```
data-client_id="{{ googleSigninClientID|attr }}"
```

Comment (Saparbek-Nagashibekov) [reply templates/login/login_core.html.twig]: Hi Brady,
made changes. please check commit 1d831fc20892c4d33307fd2913df46d5639f5df4 
Thanks

**Comment by bradymiller** [comment]
hi @Saparbek-Nagashibekov , Code looks good. Just let me know this is testing well for you and then will bring it in. thanks!

**Comment by stephenwaite** [comment]
this is testing well 

I think a separate issue is that the login process is not checking the `Google Email for Login` field in `Edit User`, `google_signin_email`, against the google email used in the oauth flow

**Comment by sjpadgett** [comment]
@kchapple or @growlingflea 
Can one of you guys help us ensure we are fixing this correctly so we can get in next patch. Thanks.

**Comment by stephenwaite** [comment]
ran another test with another email and it successfully finds the user with the google email and signs in without having to enter username/password

**Review by stephenwaite — APPROVED** [review]

**Comment by bradymiller** [comment]
Code looks good. Audited code on the server side of things and basically impossible to login unless authenticated (the passed token is confirmed with the stored client_id and google sign in email on the server end), so if the login is working on testing then this is good to bring in. Will bring this in and also bring into rel-700 branch. thanks @Saparbek-Nagashibekov !

**Comment by bradymiller** [comment]
also, thanks @stephenwaite for testing!

2DV

**Comment by stephenwaite** [comment]
southpark_cheering
